src/scf/mixing.jl

```diff
@@ -173,65 +174,39 @@ real space using a GMRES. Either the full kernel (`RPA=false`) or only the Hartr
-    χ0applies = [χ0(basis; ρin=ρin, ρ_spin_in=ρ_spin_in, kwargs...) for χ0 in mixing.χ0terms]
+    χ0applies = [χ0(basis; ρin=ρin, kwargs...) for χ0 in mixing.χ0terms]
     χ0applies = [apply for apply in χ0applies if !isnothing(apply)]
 
     # If no applies left, do not bother running GMRES and directly do simple mixing
-    isempty(χ0applies) && return mix(SimpleMixing(α=mixing.α), basis, δF_tot, δF_spin)
+    isempty(χ0applies) && return mix(SimpleMixing(α=mixing.α), basis, δF)
 
     # Solve J δρ = δF with J = (1 - χ0 vc) and χ_0 given as the sum of the χ0terms
-    devec(x) = reshape(x, size(δF_tot)..., n_spin)
-    function Jop(x)
-        δF = devec(x)  # [:, :, :, 1] is spin-up (or total), [:, :, :, 2] is spin-down
-
+    devec(x) = reshape(x, size(δF))
+    function Jop(δρ)
+        δρ = devec(δρ)
         # Apply Kernel (just vc for RPA and (vc + K_{xc}) if not RPA)
-        @views if n_spin == 1
-            x_tot  = from_real(basis, δF[:, :, :, 1])
-            x_spin = nothing
-        else
-            x_tot  = from_real(basis, δF[:, :, :, 1] .+ δF[:, :, :, 2])
-            x_spin = from_real(basis, δF[:, :, :, 1] .- δF[:, :, :, 2])
-        end
-        δV = apply_kernel(basis, x_tot, x_spin; ρ=ρin, ρspin=ρ_spin_in, RPA=mixing.RPA)
-
-        # set DC of δV to zero (δV[1] is spin-up or total, δV[2] is spin-down)
-        δV_DC = mean(mean(δV[σ].real) for σ in 1:n_spin)
-        δV[1].real .-= δV_DC
-        n_spin == 2 && (δV[2].real .-= δV_DC)
-
-        JδF = copy(δF)
+        δV = apply_kernel(basis, δρ; ρ=ρin, RPA=mixing.RPA)
+        δV .-= mean(δV)
+        Jδρ = copy(δρ)
         for apply_term! in χ0applies
-            apply_term!(JδF, δV, -1)  # JδF .-= χ0 * δV
+            apply_term!(Jδρ, δV, -1)  # JδF .-= χ0 * δV
         end
-        vec(JδF .-= mean(JδF))  # Zero DC component in total density response
+        Jδρ .-= mean(Jδρ)
+        vec(Jδρ)
     end
 
-    if n_spin == 1
-        δF_updown = δF_tot.real
-    else
-        δF_updown = cat((δF_tot.real .+ δF_spin.real) ./ 2,  # spin-up
-                        (δF_tot.real .- δF_spin.real) ./ 2,  # spin-down
-                        dims=4)
-    end
-    δF_updown .-= mean(δF_updown)  # Zero DC of δF_updown
-    J = LinearMap(Jop, length(δF_updown))
-    x = gmres(J, vec(δF_updown), verbose=mixing.verbose)
+    δF .-= mean(δF)
+    J = LinearMap(Jop, length(δF))
     # TODO Further improvement: Adapt tolerance of gmres to norm(ρ_out - ρ_in)
-
+    x = gmres(J, vec(δF), verbose=mixing.verbose)
     δρ = T(mixing.α) .* devec(x)  # Apply damping
-    δρ .+= (sum(δF_tot.real) - sum(δρ)) / length(δF_tot)  # Set DC from δF
-
-    @views if n_spin == 1
-        from_real(basis, δρ[:, :, :, 1]), nothing
-    else
-        (from_real(basis, δρ[:, :, :, 1] .+ δρ[:, :, :, 2]),
-         from_real(basis, δρ[:, :, :, 1] .- δρ[:, :, :, 2]))
-    end
+    # Set DC from δF
+    δρ .+= mean(δF)
```

Comment:
This won't work as you just removed the zero DC from δF. You need to store the DC and restore it here.

Reply:
duh, of course, thanks